### PR TITLE
add trivial constraints and constraints spotted

### PR DESCRIPTION
**Pull request by Ben-Hanan**
If you guys approve of the changes then can just merge first. I tried my best to add all the constraints i could see.

### CODE REVIEW

File: project.sql [hunk 1]
```diff
@@ -84,20 +89,21 @@ CREATE TABLE Full_time_instructors (
...
 -- DONE
 CREATE TABLE Customers (
-    cust_id                         INT UNIQUE,
-    phone                           INT,
-    address                         VARCHAR, /* added in address cuz its needed*/
-    email                           VARCHAR,
-    name                            VARCHAR,
-    number                          VARCHAR(16) UNIQUE, /** added inside unique to ensure each credit card is owned by only one customer*/
+    cust_id                         INT UNIQUE NOT NULL,
+    phone                           INT NOT NULL,
```
Comment (Ben-Hanan) [inline project.sql:102]: we can consider changing phone to another type like varchar

can also consider changing cust_id to SERIAL type or smth

File: project.sql [hunk 2]
```diff
@@ -201,3 +226,18 @@ CREATE TABLE Cancels (
...
+Constraints I think are not captured:
+
+The offerings for the same course have different launch dates.
+
+The seating capacity of a course session is equal to the seating capacity of the room where the session is conducted, 
+and the seating capacity of a course offering is equal to the sum of the seating capacities of its sessions
+
+A course offering is said to be available if the number of registrations received is no more than its seating capacity; 
+otherwise, we say that a course offering is fully booked.
+
```
Comment (Ben-Hanan) [inline project.sql:239]: list is incomplete lol didnt have time to look through everything

File: project.sql [hunk 3]
```diff
@@ -9,42 +9,47 @@ Full_time_Emp, Part_time_Emp, Employees, Pay_slips;
 
 -- DONE
 CREATE TABLE Course_packages (
-    package_id                      INT PRIMARY KEY,
-    name                            VARCHAR,
-    price                           INT, 
-    sale_start_date                 DATE,
-    sale_end_date                   DATE,
-    num_free_registrations          INT 
+    package_id                      INT PRIMARY KEY NOT NULL,
```
Comment (hanstirtaputra) [inline project.sql]: I think declaring `package_id` to be `PRIMARY KEY` already guarantees that it's `NOT NULL` right? Nonetheless if there's no harm in leaving the `NOT NULL` there then i guess its okay? @IsaacTin @thespacecuber thoughts?

File: project.sql [hunk 4]
```diff
@@ -84,20 +89,21 @@ CREATE TABLE Full_time_instructors (
...
 
 -- DONE
 CREATE TABLE Customers (
-    cust_id                         INT UNIQUE,
-    phone                           INT,
-    address                         VARCHAR, /* added in address cuz its needed*/
-    email                           VARCHAR,
-    name                            VARCHAR,
-    number                          VARCHAR(16) UNIQUE, /** added inside unique to ensure each credit card is owned by only one customer*/
+    cust_id                         INT UNIQUE NOT NULL,
```
Comment (hanstirtaputra) [inline project.sql]: Likewise for this, with the `PRIMARY KEY` already set for (cust_id, number)

File: project.sql [hunk 5]
```diff
@@ -136,23 +142,26 @@ CREATE TABLE Administrators (
...
+    FOREIGN KEY (eid) REFERENCES Administrators(eid),
+    CONSTRAINT target_number_registrations_positive CHECK (target_number_registrations >= 0),
+    CONSTRAINT fees_positive CHECK (fees >= 0),
+    CONSTRAINT registration_deadline_10_days_before_start_date CHECK (start_date - registration_deadline >= 10)
 );
 
 -- DONE
 CREATE TABLE CourseOfferingSessions (
-    sid                             INT UNIQUE,
+    sid                             INT UNIQUE, /* i rly dont think this should be unique cause 'The sessions for a course offering are numbered consecutively starting from 1'*/
```
Comment (hanstirtaputra) [inline project.sql:164]: > i rly dont think this should be unique cause 'The sessions for a course offering are numbered consecutively starting from 1 

In that case, I think so too, furthermore it's a weak entity set which means that `sid` in itself shouldn't uniquely identify `CourseOfferingSessions`. However I face some issues in Q21-24, where the input to the function/procedure only includes `course_id` and `session_id` (without `launch_date`). As such I'm not sure how to uniquely identify the `CourseOfferingSession` (See #1 for the attempted solution).

I think for now we can merge this in first and keep a note for this!

File: project.sql [hunk 6]
```diff
@@ -163,7 +172,23 @@ CREATE TABLE CourseOfferingSessions (
     FOREIGN KEY (rid) REFERENCES Rooms(rid),
     FOREIGN KEY (eid) REFERENCES Instructors(eid),
     FOREIGN KEY (course_id, launch_date) REFERENCES CourseOfferings(course_id, launch_date) ON DELETE CASCADE,
-    PRIMARY KEY (sid, launch_date, course_id) /*Weak entity of Sessions is identified by weak entity of Offering which is identified by Course*/
+    PRIMARY KEY (sid, launch_date, course_id), /*Weak entity of Sessions is identified by weak entity of Offering which is identified by Course*/
+    CONSTRAINT time_check CHECK (
+        (start_time IS NULL OR end_time IS NULL) /* done cause idk whether to put NOT NULL at start time and end time */
+        OR 
+        (CASE
```
Comment (hanstirtaputra) [inline project.sql:177]: Nice, with this I don't need to do the check in my function:)

**Review by hanstirtaputra — APPROVED** [review]
LGTM! Just some questions regarding `NOT NULL` for attributes that are already `PRIMARY KEY`.

File: project.sql (hunk 1 above)
Comment (hanstirtaputra) [reply project.sql:102]: Why `SERIAL` for `cust_id`? Is it a concern where `INT` may be insufficient to store all customers?

File: project.sql (hunk 3 above)
Comment (hanstirtaputra) [reply project.sql]: @IsaacTin @TheSpaceCuber If you guys are okay go ahead and merge in this PR!

File: project.sql (hunk 1 above)
Comment (Ben-Hanan) [reply project.sql:102]: Cause i think there's no way to know where we stopped unless we count the number of entries in our customer table and +1 every time we want to make a new id for customer I believe `SERIAL` could be a better solution as it auto increments. 

Either that or we can do something like
`cust_id INT NOT NULL DEFAULT NEXTVAL('table_name_id_seq')`

as suggested by https://www.postgresqltutorial.com/postgresql-serial/

File: project.sql (hunk 5 above)
Comment (Ben-Hanan) [reply project.sql:164]: yup agreed

File: project.sql (hunk 6 above)
Comment (Ben-Hanan) [reply project.sql:177]: can keep in mind the `(start_time IS NULL OR end_time IS NULL)`
if start_time/end_time will never be/is not supposed to be NULL then can just remove that part!

File: project.sql (hunk 3 above)
Comment (Ben-Hanan) [reply project.sql]: right usually i leave the primary key alone i think i just auto added haha mb!

File: project.sql (hunk 4 above)
Comment (Ben-Hanan) [reply project.sql]: yup same as above

File: project.sql (hunk 1 above)
Comment (IsaacTin) [reply project.sql:102]: Just researched it up. I think we should use serial for all primary key ids

File: project.sql [hunk 7]
```diff
@@ -163,7 +172,23 @@ CREATE TABLE CourseOfferingSessions (
...
+        OR 
+        (CASE 
+            WHEN (start_time >= TIME '09:00:00' AND end_time <= TIME '18:00:00' AND start_time <> end_time) THEN
+                CASE
+                    WHEN (start_time >= TIME '09:00:00' AND start_time <= TIME '12:00:00') THEN (end_time <= TIME '12:00:00')
+                    WHEN (end_time <= TIME '16:00:00' AND end_time >= TIME '14:00:00') THEN (start_time >= TIME '14:00:00')
+                    ELSE FALSE
+                END
+            ELSE FALSE
+        END
```
Comment (IsaacTin) [inline project.sql]: Maybe using BETWEEN is better? like IF start_time BETWEEN 12 to 2 OR end_time BETWEEN 12 - 2

File: project.sql (hunk 6 above)
Comment (IsaacTin) [reply project.sql:177]: I think it should never be null

File: project.sql (hunk 1 above)
Comment (TheSpaceCuber) [reply project.sql:102]: https://www.postgresqltutorial.com/postgresql-identity-column/

Here it says `SERIAL` is not SQL compliant and there's a new way `GENERATED AS IDENTITY`.

`create table t1 (id serial primary key);`
vs
`create table t2 (id integer primary key generated always as identity);`

**Comment by TheSpaceCuber** [comment]
LGTM

**Pull request merged**